File: dynamixel_workbench_practice/src/master_slave_service.cpp
#include "dynamixel_workbench_practice/master_slave_service.h"

void jointStateCallback(const dynamixel_workbench_msgs::DynamixelStateListConstPtr& msg)
{

  srv.request.command = "";
  srv.request.id = 2;
  srv.request.addr_name = "Goal_Position";
  srv.request.value = msg->dynamixel_state[0].present_position;

  if (client.call(srv))
    {
      // ROS_INFO("True (=1) or False (=0) : %d", srv.response.comm_result);
    }
  else
    {
      ROS_ERROR("Failed to call service.");
      return;
    }

  // loop rate regulation
  // ros::Rate loop_rate(control_cycle);
  // loop_rate.sleep(); // wait for dynamixel_command service.
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "dynamixel_workbench_master_slave_demo");

  ros::NodeHandle nh;
  client = nh.serviceClient<dynamixel_workbench_msgs::DynamixelCommand>("/dynamixel_workbench/dynamixel_command");

  srv.request.command = "";
  srv.request.id = 1;
  srv.request.addr_name = "Torque_Enable";
  srv.request.value = 0;

  if (client.call(srv))
    {
      ROS_INFO("True (=1) or False (=0) : %d", srv.response.comm_result);
    }
  else
    {
      ROS_ERROR("Failed to call service.");
      return 1;
    }

  ros::spinOnce(); // wait for the service request above

  ros::Subscriber joint_sub = nh.subscribe("/dynamixel_workbench/dynamixel_state", queue, jointStateCallback);

  // infinite loop w/o rate
  ros::spin();

  // infinite loop w/ rate
  // ros::Rate loop_rate(control_cycle);
  // while (ros::ok())
  //   {
  //     ros::spinOnce();
  //     loop_rate.sleep();
  //   }

  return 0;
}

File: dynamixel_workbench_practice/scripts/master_slave_service.py
#!/usr/bin/env python
import rospy
# from dynamixel_workbench_msgs.msg import DynamixelState
from dynamixel_workbench_msgs.msg import DynamixelStateList
from dynamixel_workbench_msgs.srv import DynamixelCommand
from dynamixel_workbench_msgs.srv import DynamixelCommandResponse

control_cycle = 125 # use if willing to regulate control cycle.
queue = 1
id_1 = 7 # 1
id_2 = 8 # 2

# subscribe joint states and call dynamixel command service.
def joint_states_callback(msg):

    # print "current master angle: %s" % msg.dynamixel_state[0].present_position

    # service clinent implementation
    rospy.wait_for_service('dynamixel_workbench/dynamixel_command')
    try:
        service = rospy.ServiceProxy('dynamixel_workbench/dynamixel_command', DynamixelCommand)
        response = service('', id_2, 'Goal_Position', msg.dynamixel_state[0].present_position)
    except rospy.ServiceException, e:
        print "Service call failed: %s" % e
        return DynamixelCommandResponse

    # loop w/ rate
    # rate = rospy.Rate(control_cycle)
    # rate.sleep()

def command_client():

    # initialize node
    rospy.init_node('dynamixel_command_client')

    # turn off torque of master (id=1)
    rospy.wait_for_service('dynamixel_workbench/dynamixel_command')

    try:
        service = rospy.ServiceProxy('dynamixel_workbench/dynamixel_command', DynamixelCommand)
        response = service('', id_1, 'Torque_Enable', 0)
        print response.comm_result
    except rospy.ServiceException, e:
        print "Service call failed: %s" % e
        return DynamixelCommandResponse

    # subscriber
    joint_states_sub = rospy.Subscriber('/dynamixel_workbench/dynamixel_state', DynamixelStateList,
                                        joint_states_callback, queue_size=queue)

    # ROS loop w/o rate
    rospy.spin()

    # ROS loop w/ rate
    # rate = rospy.Rate(control_cycle)
    # while not rospy.is_shutdown():
    #     rate.sleep()

if __name__ == "__main__":
    command_client()
